jscomp/frontend/ast_external_process.ml

```diff
@@ -492,12 +492,12 @@ let process_obj
                    param_type::arg_types, result_types
                  | Nothing ->
                    let s = (Lam_methname.translate  name) in
-                   let for_sure_not_nested =
-                     match ty.ptyp_desc with
-                     | Ptyp_constr({txt = Lident txt;_}, []) ->
-                       Ast_core_type.is_builtin_rank0_type txt
-                     | _ -> false in
-                   {obj_arg_label = External_arg_spec.optional for_sure_not_nested s; obj_arg_type},
+                   (* XXX(anmonteiro): it's unsafe to just read the type of the
+                      labelled argument declaration, since it could be `'a` in
+                      the implementation, and e.g. `bool` in the interface. See
+                      https://github.com/anmonteiro/bucklescript/pull/58 for
+                      a test case. *)
+                   {obj_arg_label = External_arg_spec.optional false s; obj_arg_type},
```

Comment:
As mentioned in off, I'm pretty sure this kind of "type driven optimization" will bite us in the future, like it happened here.

Reply:
I have actually been thinking about this, especially the way `option<'a>` is special cased 
And why does Nothing in `type Maybe('a) = Something('a) | Nothing;` gets optimized to an integer...? I feel like with how V8 inlines things when the types match that's doing a _deoptimization_ but I haven't had the chance to run any benchmarks

Reply:
@Mathspy Yup, I think the option specifically it is okay as it's a common pattern, but it's been a long time and nowadays turbofan is less sensitive to that, but definitely mixing primitives is bad for performance.

But we don't have an alternative, it's an union type after all. But I'm not sure what is easy for engines to optimize, `{ kind: 1 }` and `{ kind: 1, data: "lol" }` or `1` and `{ kind: 1, data: "lol" }`

Reply:
Also nowadays there is ignition besides of turbofan, which does a lot less of specialization

Reply:
Ah, that’s good to know about Ignition 
And yeah I just find the option optimization to awkwardly have a bit of runtime cost once you start using generics which seems counterintuitive for an “optimization”